### PR TITLE
Add webpack to bundle selfrepair assets

### PR DESCRIPTION
**Pull request by brittanystoroz**
@mythmon r? 

You can run `./node_modules/.bin/webpack --config ./normandy/webpack.config.js` to create the bundle manually. `--watch`. I'm not 100% sure if there's anything particular we have to do to set this up for the production environment? We can a prod-specific webpack config if there are significant changes we need to make.

### CODE REVIEW

**Comment by mythmon** [comment]
This error is popping up during tests: "Error reading webpack-stats.json. Are you sure webpack has generated the file and the path is correct?". I wonder if we have to generate a webpack bundle before running the tests?

**Comment by brittanystoroz** [comment]
Test failures are related to not finding the webpack-stats.json file...which I've .gitignored locally. Is there a place for me to add the webpack command so it'll run as part of the build step...?

**Comment by brittanystoroz** [comment]
@mythmon yea it seems like that is the case, I was reading https://github.com/owais/django-webpack-loader#how-to-use-in-production Just not sure where in our project we add something like that

**Comment by mythmon** [comment]
Will we need `webpack-stat.json` in production? If so, we could add a webpack line to the docker file. I think right before we run `collectstatic`. If we don't need it in production, we can add it as a dependencies command in `circle.yml`.

File: normandy/webpack.config.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+var path = require("path");
```
Comment (mythmon) [inline normandy/webpack.config.js]: Nit: We should be consistent about quotes.

File: normandy/selfrepair/static/js/index.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+import 'babel-polyfill'
+import './normandy_driver.js'
+import './self_repair_runner.js'
```
Comment (mythmon) [inline normandy/selfrepair/static/js/index.js]: Why import `./normandy_driver.js` here? Wouldn't it get included through `self_repair_runner.js`?

File: normandy/selfrepair/templates/selfrepair/repair.html [hunk 3]
```diff
@@ -7,13 +8,6 @@
...
   <body>
-    <script src="{% static 'js/uitour.js' %}"></script>
-    <script src="{% static 'npm/node-uuid/uuid.js' %}"></script>
-    <script src="{% static 'npm/wolfy87-eventemitter/EventEmitter.js' %}"></script>
-    <script src="{% static 'npm/babel-polyfill/dist/polyfill.js' %}"></script>
-    <!-- Cheat so that we get let and other version-limited JavaScript features. -->
-    <script type="application/javascript;version=1.8" src="{% static 'js/xhr.js' %}"></script>
-    <script type="application/javascript;version=1.8" src="{% static 'js/normandy_driver.js' %}"></script>
-    <script type="application/javascript;version=1.8" src="{% static 'js/self_repair_runner.js' %}"></script>
+    {% render_bundle 'main' %}
```
Comment (mythmon) [inline normandy/selfrepair/templates/selfrepair/repair.html]: Where does the name `main` come from?

**Comment by mythmon** [comment]
This looks nice and simple. Way better than the devserver idea we talked about way back when. I'm going to try and run it locally now.

**Comment by mythmon** [comment]
When running `npm install`, I get

```
npm WARN babel-loader@6.2.4 requires a peer of babel-core@^6.0.0 but none was installed.
npm WARN babel-loader@6.2.4 requires a peer of webpack@1 || ^2.1.0-beta but none was installed.
```

**Comment by brittanystoroz** [comment]
@mythmon  ahh what version of npm are you using? I might be on an old version, npm3 no longer installs peers for you. Will fix.

AFAIK you shouldn't actually need webpack-stats.json in prod.

File: normandy/selfrepair/templates/selfrepair/repair.html (hunk 3 above)
Comment (brittanystoroz) [inline normandy/selfrepair/templates/selfrepair/repair.html]: This is just the default bundle reference name, I believe I can override it in the webpack config with something more specific.

**Comment by mythmon** [comment]
```shell
$ npm --version
3.8.7
```

Also, I think `babel-preset-es2015` is missing.

**Comment by mythmon** [comment]
Good news, after installing the missing bits abpve, this worked great. When I visited https://localhost:8000/en-US/repair/, I got exactly one request for the bundle. :fireworks: :fireworks:

File: normandy/webpack.config.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
+var path = require('path');
```
Comment (mythmon) [inline normandy/webpack.config.js]: Most projects I've seen put this at the root of the git repo. Is there a reason to put it in the `normandy` subdirectory here?

**Comment by mythmon** [comment]
After reading the django-webpack-loader "How to use in production" section, I think adding a call to webpack to `Dockerfile` is the right way to go, so we get a `webpack-stats.json` in the image.

I made this change to `Dockerfile`, and now running the server in Docker just works (I used Docker Compose and [normandy-compose](https://github.com/mythmon/normandy-compose) to do that). I suspect this will make the tests pass too.

```diff
diff --git a/Dockerfile b/Dockerfile
index 7f6a9da..0cef5c2 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -11,10 +11,14 @@ RUN curl -s https://deb.nodesource.com/gpgkey/nodesource.gpg.key | apt-key add -
     apt-get update && apt-get install -y nodejs
 
 COPY ./requirements.txt /app/requirements.txt
-RUN pip install -U 'pip>=8' && pip install --upgrade --no-cache-dir -r requirements.txt
+COPY ./package.json /app/package.json
+RUN pip install -U 'pip>=8' && \
+    pip install --upgrade --no-cache-dir -r requirements.txt && \
+    npm install
 
 COPY . /app
-RUN DJANGO_CONFIGURATION=Build ./manage.py collectstatic --no-input && \
+RUN ./node_modules/.bin/webpack && \
+    DJANGO_CONFIGURATION=Build ./manage.py collectstatic --no-input && \
     mkdir -p media && chown app:app media && \
     mkdir -p __version__ && \
     git rev-parse HEAD > __version__/commit && \
```

**Comment by brittanystoroz** [comment]
🎉

File: normandy/selfrepair/static/js/self_repair_runner.js [hunk 5]
```diff
@@ -1,3 +1,8 @@
+import 'babel-polyfill'
+import xhr from './xhr.js'
+import Mozilla from './uitour.js'
+import Normandy from './normandy_driver.js'
```
Comment (mythmon) [inline normandy/selfrepair/static/js/self_repair_runner.js]: This file needs uuid imported too. I only get an error for this when I run this in docker-compose, not when I run it normally, which is weird.

Comment (mythmon) [inline normandy/selfrepair/static/js/self_repair_runner.js]: Oh! When I ran it normally, I used `https://normandy/` (my normal test domain). When I tested it with docker-compose, I used `https://localhost:8000`, which doesn't share local storage. This module only needs `uuid` once, and then it caches that value in localstorage, which is per domain. That explains it.

**Comment by mythmon** [comment]
Woot, nice job! One more tiny issue with this (missing an import), and I think this is good to go.

Well, maybe we want to add that bit about `webpack` and/or `webpack --watch` somewhere? I don't know about that.

I've got to step out for a while, but I'm r+ after the import statement is fixed.

**Comment by mythmon** [comment]
Nice. This all works for me now.

**Pull request merged**